internal/webhooks/vmoperator/vspheremachine.go

```diff
@@ -0,0 +1,87 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-vsphere/internal/webhooks"
+)
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-vmware-infrastructure-cluster-x-k8s-io-v1beta1-vspheremachine,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachines,versions=v1beta1,name=validation.vspheremachine.vmware.infrastructure.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-vmware-infrastructure-cluster-x-k8s-io-v1beta1-vspheremachine,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachines,versions=v1beta1,name=default.vspheremachine.vmware.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+// VSphereMachineWebhook implements a validation and defaulting webhook for vm-operator VSphereMachine.
+type VSphereMachineWebhook struct{}
+
+var _ webhook.CustomValidator = &VSphereMachineWebhook{}
+var _ webhook.CustomDefaulter = &VSphereMachineWebhook{}
+
+func (webhook *VSphereMachineWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&vmwarev1.VSphereMachine{}).
+		WithValidator(webhook).
+		WithDefaulter(webhook).
+		Complete()
+}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (webhook *VSphereMachineWebhook) Default(_ context.Context, _ runtime.Object) error {
+	return nil
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *VSphereMachineWebhook) ValidateCreate(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *VSphereMachineWebhook) ValidateUpdate(_ context.Context, oldRaw runtime.Object, newRaw runtime.Object) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+
+	newTyped, ok := newRaw.(*vmwarev1.VSphereMachine)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a vm-operator VSphereMachine but got a %T", newRaw))
+	}
+
+	oldTyped, ok := oldRaw.(*vmwarev1.VSphereMachine)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a vm-operator VSphereMachine but got a %T", oldRaw))
+	}
+
+	newSpec, oldSpec := newTyped.Spec, oldTyped.Spec
+	if newSpec.ImageName != oldSpec.ImageName || newSpec.ClassName != oldSpec.ClassName || newSpec.StorageClass != oldSpec.StorageClass || newSpec.MinHardwareVersion != oldSpec.MinHardwareVersion {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), newTyped.Spec, "updating ImageName, ClassName, StorageClass or MinHardwareVersion is not allowed"))
+	}
```

Comment:
Not sure if we should have all these in seperate ifs instead. WDYT @sbueringer ?

Maybe worth adding a comment that these are immutable on the vm operator api side.

Reply:
Make sense to me, that should make it more readable :-)